File: Sources/WickTrading/PositionEntryPlanner.swift
import CryptoKit
import Foundation

/// Decides which position days still need a journal item for display, and
/// which symbols those new items should carry.
///
/// Existing tags are never rewritten. A position is covered when any item on
/// its opening day already matches the symbol; otherwise one item per missing
/// distinct symbol is planned, whether or not the day entry already exists.
public enum PositionEntryPlanner {
    public struct PlannedDay: Equatable, Sendable {
        /// Start of the local calendar day.
        public var day: Date
        public var dayKey: String
        /// Distinct symbols opened that day, sorted - one item per symbol.
        public var symbols: [String]
    }

    public static func plan(
        positions: [TradingPosition],
        existingTagsByDay: [String: [String]],
        dayKey: (Date) -> String,
        startOfDay: (Date) -> Date
    ) -> [PlannedDay] {
        var byDay: [String: (day: Date, symbols: Set<String>)] = [:]

        for position in positions {
            let key = dayKey(position.openTime)
            let existingTags = existingTagsByDay[key] ?? []
            guard !existingTags.contains(where: {
                SymbolTagMatcher.matches(tag: $0, symbol: position.symbol)
            }) else { continue }

            if byDay[key] == nil {
                byDay[key] = (startOfDay(position.openTime), [])
            }
            byDay[key]!.symbols.insert(position.symbol)
        }

        return byDay
            .map { key, group in
                PlannedDay(
                    day: group.day,
                    dayKey: key,
                    symbols: group.symbols.sorted()
                )
            }
            .sorted { $0.day < $1.day }
    }

    /// Stable across devices so concurrent first syncs create the same item
    /// identity. Dropbox can then merge the two day snapshots without keeping
    /// duplicate tag-only items that would render every position twice.
    public static func stableItemID(
        journalID: UUID,
        dayKey: String,
        symbol: String
    ) -> UUID {
        let components = [
            "wick.exchange-item.v1",
            journalID.uuidString.lowercased(),
            dayKey,
            symbol.trimmingCharacters(in: .whitespacesAndNewlines).uppercased(),
        ]
        let source = components.joined(separator: "|")
        var bytes = Array(SHA256.hash(data: Data(source.utf8)).prefix(16))
        // Mark the digest as an RFC 4122 name-based UUID.
        bytes[6] = (bytes[6] & 0x0f) | 0x50
        bytes[8] = (bytes[8] & 0x3f) | 0x80
        return UUID(uuid: (
            bytes[0], bytes[1], bytes[2], bytes[3],
            bytes[4], bytes[5], bytes[6], bytes[7],
            bytes[8], bytes[9], bytes[10], bytes[11],
            bytes[12], bytes[13], bytes[14], bytes[15]
        ))
    }
}

File: linux/src/trading/PositionEntryPlanner.cpp
#include "PositionEntryPlanner.h"

#include "Crypto.h"
#include "SymbolTagMatcher.h"

#include <algorithm>
#include <cctype>
#include <chrono>
#include <cstring>
#include <set>

namespace wick {

std::vector<PlannedPositionDay> PositionEntryPlanner::plan(
    const std::vector<TradingPosition> &positions,
    const std::map<std::string, std::vector<std::string>> &existingTagsByDay,
    const std::function<std::string(TimePoint)> &dayKey,
    const std::function<TimePoint(TimePoint)> &startOfDay)
{
    struct Group {
        TimePoint day{};
        std::set<std::string> symbols;
    };
    std::map<std::string, Group> byDay;
    for (const auto &position : positions) {
        const TimePoint open{std::chrono::milliseconds(position.openTime)};
        const std::string key = dayKey(open);
        const auto itTags = existingTagsByDay.find(key);
        const std::vector<std::string> existing =
            itTags == existingTagsByDay.end() ? std::vector<std::string>{} : itTags->second;
        bool covered = false;
        for (const auto &tag : existing) {
            if (SymbolTagMatcher::matches(tag, position.symbol)) {
                covered = true;
                break;
            }
        }
        if (covered)
            continue;
        auto &group = byDay[key];
        if (group.symbols.empty())
            group.day = startOfDay(open);
        group.symbols.insert(position.symbol);
    }

    std::vector<PlannedPositionDay> out;
    out.reserve(byDay.size());
    for (auto &[key, group] : byDay) {
        PlannedPositionDay row;
        row.day = group.day;
        row.dayKey = key;
        row.symbols.assign(group.symbols.begin(), group.symbols.end());
        std::sort(row.symbols.begin(), row.symbols.end());
        out.push_back(std::move(row));
    }
    std::sort(out.begin(), out.end(),
              [](const PlannedPositionDay &a, const PlannedPositionDay &b) { return a.day < b.day; });
    return out;
}

Uuid PositionEntryPlanner::stableItemID(const Uuid &journalID, std::string_view dayKey,
                                        std::string_view symbol)
{
    std::string upper(symbol);
    for (char &c : upper)
        c = static_cast<char>(std::toupper(static_cast<unsigned char>(c)));
    const std::string source = std::string("wick.exchange-item.v1|")
        + journalID.toLowerString() + "|" + std::string(dayKey) + "|" + trimCopy(upper);
    const auto digest = sha256Bytes(source);
    Uuid u;
    std::memcpy(u.bytes.data(), digest.data(), 16);
    u.bytes[6] = static_cast<uint8_t>((u.bytes[6] & 0x0f) | 0x50);
    u.bytes[8] = static_cast<uint8_t>((u.bytes[8] & 0x3f) | 0x80);
    return u;
}

} // namespace wick
